tests/library/test_type_utils.py

```diff
@@ -262,3 +263,15 @@ def test_parse_unsupported_type(self):
     def test_parse_malformed_string(self):
         with self.assertRaises(TypeError):
             parse_type_string("List[[int]]")
+
+    def test_format_type_string(self):
+        self.assertEqual(
+            format_type_string(
+                "typing.Tuple[int,float,int|list[int|dict[str,int|float]]]"
+            ),
+            "Tuple[int,float,Union[int,List[Union[int,Dict[str,Union[int,float]]]]]]",
+        )
+        self.assertEqual(
```

Comment:
Hi @pawelknes, I am confused: from the first assert, I learn that you want to convert `typing.Tuple` into `Tuple`, whereas from the second: `tuple` into typing.Tuple`.  Why, then, stop there, and not convert `tuple` into `Tuple'?

Reply:
The first assert just checks if `format_type_string` properly formats a type string so that it is compatible with `parse_type_string` function which requires strings in such format. 

While the second one performs the entire pipeline of parsing a type string into a type hint; `typing.Tuple[int, str, typing.List[typing.Union[int, float]]]` is just a type hint and is checked against output of `parse_type_string("tuple[int, str, typing.List[int | float]]")` to ensure if parsing works correctly for such type strings. It could be written as just `Tuple` without `typing.` but this is the same

Reply:
I see. thanks.    Do you flatten the Union-s?  e.g. Union inside Union?  or Union of some types, two of which are identical?   Or just transform from | to Union?

Reply:
```
format_type_string(
                "Tuple[int,float,int|list[int|dict[str,int|float]]] | bool | str")  => 
```
                
`Union[Tuple[int,float,Union[int,List[Union[int,Dict[str,Union[int,float]]]]]],Union[bool,str]]`

I think that Union should not be used inside Union. I.e., I think that here, the expected result should be:
`Union[Tuple[int,float,Union[int,List[Union[int,Dict[str,Union[int,float]]]]]],bool,str]`

Reply:
It just transforms a string, it doesn't simplify it in any way since that would probably require a different approach

Reply:
> ```
> format_type_string(
>                 "Tuple[int,float,int|list[int|dict[str,int|float]]] | bool | str")  => 
> ```
> 
> `Union[Tuple[int,float,Union[int,List[Union[int,Dict[str,Union[int,float]]]]]],Union[bool,str]]`
> 
> I think that Union should not be used inside Union. I.e., I think that here, the expected result should be: `Union[Tuple[int,float,Union[int,List[Union[int,Dict[str,Union[int,float]]]]]],bool,str]`

You're right. I didn't test it for multiple options in a single union. I'll make a fix

Reply:
I thought about this and there is no need to parse a type hint further on a string level. It'll be done by `parse_type_string` function anyway; only type string needs to be properly formatted and that is done by the function (since `Union[a, b, c]`  is basically equivalent to `Union[a, Union[b, c]]`)

Reply:
I see, so if format_type_string is only used as a service for parse_type_string , then its alright.

What do you do if the | sign shows inside a string, being a member of typing.Literal?

`print(format_type_string( "typing.Tuple[int,float,Literal[\"lef|t\"],int|list[int|dict[str,int|float]]]" ))`
prints:
`Tuple[int,float,Literal[Union["lef,t"]],Union[int,List[Union[int,Dict[str,Union[int,float]]]]]]`

Reply:
I think that your code can not support these surprising Literals, 
However, the current `parse_type_string` in unitxt does not support Literal at all.
I am not sure there is a way to convert Literal (as you used in your code, for example: `Literal["left", "right"]`) to anything digestible by `parse_type_string` that is currently in unitxt.

So I think this should be made clearer in your documentation: it does not 
`"Formats a string representing a Python type hint so that it is compatible with any Python version (before or after Python 3.10)."`

such that the result (of your code) is correctly digestible by `parse_type_string` currently in unitxt.

I think that indeed, as you explained to me, your code can deal with all "residents" of typing of all python levels, because your code only depends on the [ ],  but this is correct for all residents other than resident "Literal".

Reply:
Good point, I didn't consider that. I could add some additional rules for formatting insides of Literal, but since unitxt doesn't support it, I'll just exclude it and write that Literal isn't supported